tool/public/client.js

```diff
@@ -9,28 +9,45 @@ class ConverterUi {
 
 		iptFile.addEventListener("change", () => {
 			const reader = new FileReader();
-			reader.onload = () => iptText.value = reader.result;
 			reader.readAsText(iptFile.files[0]);
-			doConvert();
+			reader.onload = () => {
+				let text;
+
+				// Check whether filetype is legal
+				if (iptFile.value.match(/\.(json|txt)$/i) || iptFile.value.match(/$(?<!\.\w+)/)) { // .json, .txt, or no filetype
```

Comment:
`$(?<!\.\w+)` means match anything that doesn't end in `\.\w+`, i.e. without a filetype. The actual ordering of the lookbehind and end-string is arbitrary, since they both match character _boundaries_: `$(?<!\.\w+)` means "jump to the end of the string and check that this group doesn't precede it", while `(?<!\.\w+)$` is "for each boundary between characters, check that this group doesn't precede it and that nothing else succeeds it". (There might be some weird regex engine mess that compiles the expression to the same code; I just find the former more intuitive.)

Reply:
I see. Either way, is this not more easily expressed as `!iptFile.value.includes(".")` ?

Reply:
... yes

I think what I was initially doing was trying to allow `.` in the filename, e.g. something like `very.long.filename` that is technically 'filetype-less'. So initially it was `\w{2,4}` or something, but then I ran into enough counterexamples amongst my own local files that I dropped the strict range, not realising what you said. Your code review is greatly appreciated 🙏